presto-common/src/main/java/com/facebook/presto/common/type/UserDefinedDistinctType.java

```diff
@@ -0,0 +1,115 @@
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class UserDefinedDistinctType
+{
+    private final QualifiedObjectName name;
+    private final TypeSignature base;
+    private final Optional<QualifiedObjectName> superTypeName;
+    private Optional<UserDefinedDistinctType> superType;
+    private final boolean isOrderable;
+
+    @JsonCreator
+    public UserDefinedDistinctType(
+            @JsonProperty("name") QualifiedObjectName name,
+            @JsonProperty("base") TypeSignature base,
+            @JsonProperty("superTypeName") Optional<QualifiedObjectName> superTypeName,
+            @JsonProperty("superType") Optional<UserDefinedDistinctType> superType,
+            @JsonProperty("isOrderable") boolean isOrderable)
+    {
+        this.name = requireNonNull(name, "name is null");
+        this.base = requireNonNull(base, "base is null");
+        this.superTypeName = requireNonNull(superTypeName, "superTypeName is null");
+        this.superType = requireNonNull(superType, "superType is null");
+        this.isOrderable = isOrderable;
+    }
+
+    public UserDefinedDistinctType withoutSuperTypeInfo()
+    {
+        return new UserDefinedDistinctType(name, base, superTypeName, Optional.empty(), isOrderable);
+    }
+
+    @JsonProperty
+    public QualifiedObjectName getName()
+    {
+        return name;
+    }
+
+    @JsonProperty
+    public TypeSignature getBase()
+    {
+        return base;
+    }
+
+    @JsonProperty
+    public Optional<UserDefinedDistinctType> getSuperType()
+    {
+        return superType;
+    }
+
+    @JsonProperty
+    public Optional<QualifiedObjectName> getSuperTypeName()
+    {
+        return superTypeName;
+    }
+
+    @JsonProperty
+    public boolean isOrderable()
+    {
+        return isOrderable;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        UserDefinedDistinctType other = (UserDefinedDistinctType) o;
+
+        return Objects.equals(name, other.name) && Objects.equals(base, other.base) && Objects.equals(superTypeName, other.superTypeName) && Objects.equals(isOrderable, other.isOrderable);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(name, base, superTypeName, isOrderable);
+    }
+
+    @Override
+    public String toString()
+    {
+        return format(
+                "%s{%s, %s, %s}",
+                name.toString(),
+                base.toString(),
+                isOrderable,
+                superType.isPresent() ? superType.get().toString() : superTypeName.map(t -> t.toString()).orElse("null"));
```

Comment:
- s/`t -> t.toString()`/`QualifiedObjectName::toString`
- maybe replace null with "N/A" because it's just we do not know the parent while it could exist